### PR TITLE
feat: revisit telemetry type filter usage outside Azure Application Insights Serilog sink

### LINKED ISSUE #206: Provide support for filtering based on telemetry type with all sinks
**Is your feature request related to a problem? Please describe.**
Today, it seems like you can only filter for Azure Application Insights while I would like to do this for console logging.

It might be just me though so just checking but based on the code I think it doesn't work.

**Describe the solution you'd like**
Provide support for filtering based on telemetry type with all sinks, instead of just Azure Application Insights.

image

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by stijnmoreels**
Updates the `TelemetryTypeFilter` so it can be used with the new telemetry log format + outside Azure Application Insights sinks.

Closes #206

### CODE REVIEW

File: src/Arcus.Observability.Telemetry.Core/Logging/DependencyLogEntry.cs [hunk 1]
```diff
@@ -49,6 +49,7 @@ public class DependencyLogEntry
             StartTime = startTime.ToString(FormatSpecifiers.InvariantTimestampFormat);
             Duration = duration;
             Context = context;
+            Context[ContextProperties.General.TelemetryType] = TelemetryType.Dependency;
```
Comment (stijnmoreels) [inline src/Arcus.Observability.Telemetry.Core/Logging/DependencyLogEntry.cs:52]: Added this, not for functionality, but because it seemed something good to have, since when we format to JSON, there's otherwise no single property that can be used to determine which type it is.
But again, it's not being used in our functionality. I'll remove this if you think this is unnecessary.

File: src/Arcus.Observability.Telemetry.Serilog.Filters/TelemetryTypeFilter.cs [hunk 2]
```diff
@@ -44,27 +48,36 @@ public static TelemetryTypeFilter On(TelemetryType telemetryType)
...
+                case TelemetryType.Dependency:
+                    return logEvent.Properties.ContainsKey(ContextProperties.DependencyTracking.DependencyLogEntry);
+                case TelemetryType.Request:
+                    return logEvent.Properties.ContainsKey(ContextProperties.RequestTracking.RequestLogEntry);
+                case TelemetryType.Events:
+                    return logEvent.Properties.ContainsKey(ContextProperties.EventTracking.EventLogEntry);
+                case TelemetryType.Metrics:
+                    return logEvent.Properties.ContainsKey(ContextProperties.MetricTracking.MetricLogEntry);
+                default:
+                    return false;
```
Comment (stijnmoreels) [inline src/Arcus.Observability.Telemetry.Serilog.Filters/TelemetryTypeFilter.cs:79]: Problem here: the `TelemetryType` also includes `Trace`, but since we don't do traces ourselves, everything outside these cases is considered being a trace. Not sure if I should remove the enum value or add a case here and always return `false`.

File: src/Arcus.Observability.Tests.Integration/Serilog/TelemetryTypeFilterTests.cs [hunk 3]
```diff
@@ -0,0 +1,1266 @@
...
+using ILogger = Microsoft.Extensions.Logging.ILogger;
+
+namespace Arcus.Observability.Tests.Integration.Serilog
+{
+    [Trait("Category", "Integration")]
+    public class TelemetryTypeFilterTests
+    {
+        private readonly Faker _bogusGenerator = new Faker();
+
+        public static IEnumerable<object[]> TelemetryTypesWithoutEvent => GetTelemetryTypesWithout(TelemetryType.Events);
```
Comment (stijnmoreels) [inline src/Arcus.Observability.Tests.Integration/Serilog/TelemetryTypeFilterTests.cs:28]: Oh, it really itched to do some property-based testing... 😄

**Comment by stijnmoreels** [comment]
I'll add @fgheysels from now on too on each PR, as he's promoted to CodeOwner?

File: src/Arcus.Observability.Telemetry.Core/Logging/DependencyLogEntry.cs (hunk 1 above)
Comment (fgheysels) [reply src/Arcus.Observability.Telemetry.Core/Logging/DependencyLogEntry.cs:52]: I currently have no opinion on that. :)  If I see the result, I can maybe form an opinion :)

File: src/Arcus.Observability.Telemetry.Serilog.Filters/TelemetryTypeFilter.cs [hunk 4]
```diff
@@ -44,27 +48,36 @@ public static TelemetryTypeFilter On(TelemetryType telemetryType)
...
+                new ArgumentOutOfRangeException(nameof(telemetryType), telemetryType, "Requires a type of telemetry that's within the supported value range of the enumeration"));
+            
             return new TelemetryTypeFilter(telemetryType, isTrackingEnabled);
         }
 
         /// <inheritdoc />
         public bool IsEnabled(LogEvent logEvent)
         {
-            if (IsTrackingEnabled == true)
+            if (IsTrackingEnabled is true)
```
Comment (fgheysels) [inline src/Arcus.Observability.Telemetry.Serilog.Filters/TelemetryTypeFilter.cs:63]: What's the advantage of writing `is true` instead of `== true` ?

File: src/Arcus.Observability.Telemetry.Serilog.Filters/TelemetryTypeFilter.cs (hunk 2 above)
Comment (fgheysels) [reply src/Arcus.Observability.Telemetry.Serilog.Filters/TelemetryTypeFilter.cs:79]: When is this method called ?  I would assume that it's called to determine if the event must be logged or not ?  Isnt that done based on the configured loglevel ?  Or am I talking rubbish here ?

**Review by fgheysels — COMMENTED** [review]
Is the way events / metrics / etc.... are written now being changed, since I haven't seen any changes on that, or did I miss something ?

File: src/Arcus.Observability.Telemetry.Serilog.Filters/TelemetryTypeFilter.cs (hunk 4 above)
Comment (stijnmoreels) [reply src/Arcus.Observability.Telemetry.Serilog.Filters/TelemetryTypeFilter.cs:63]: `is` forces the terms to be constants l, that's also why `... is null` is much safer than `... == null` bc the `==` can be overwritten. It's a personal choice maybe or my paranoia speaking. I tend to use `is` where possible.

File: src/Arcus.Observability.Telemetry.Core/Logging/DependencyLogEntry.cs (hunk 1 above)
Comment (tomkerkhove) [reply src/Arcus.Observability.Telemetry.Core/Logging/DependencyLogEntry.cs:52]: This is fine for me, as long as the Application Insights sink removes this for that telemetry, other sinks can have this without a problem imo

File: src/Arcus.Observability.Telemetry.Core/TelemetryType.cs [hunk 5]
```diff
@@ -1,14 +1,36 @@
...
+        /// </summary>
+        Request = 2,
+        
+        /// <summary>
+        /// Specifies the type of logged telemetry as custom events.
+        /// </summary>
+        Events = 4,
+        
+        /// <summary>
+        /// Specifies the type of logged telemetry as value metrics.
```
Comment (tomkerkhove) [inline src/Arcus.Observability.Telemetry.Core/TelemetryType.cs:32]:
```suggestion
        /// Specifies the type of logged telemetry as metrics.
```

File: src/Arcus.Observability.Telemetry.Core/TelemetryType.cs [hunk 6]
```diff
@@ -1,14 +1,36 @@
-﻿namespace Arcus.Observability.Telemetry.Core
+﻿using Microsoft.Extensions.Logging;
+
+namespace Arcus.Observability.Telemetry.Core
 {
     /// <summary>
-    ///     Different types of telemetry
+    /// Represents the supported types of telemetry available to track during extensions on the <see cref="ILogger"/>.
+    /// Compatible with Azure Application Insights.
```
Comment (tomkerkhove) [inline src/Arcus.Observability.Telemetry.Core/TelemetryType.cs:7]: It's compatible with more than Azure Application Insights, no? Maybe rephrase it so that we make it easier to use and add a link to our sink doc?

File: src/Arcus.Observability.Tests.Unit/Telemetry/TelemetryTypeFilterTests.cs [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        [InlineData(TelemetryType.Metrics | TelemetryType.Request | TelemetryType.Events)]
+        public void CreateFilter_WithTelemetryTypeOutsideEnum_Fails(TelemetryType telemetryType)
+        {
+            Assert.ThrowsAny<ArgumentException>(() => TelemetryTypeFilter.On(telemetryType));
+        }
+        
+        [Theory]
+        [InlineData(TelemetryType.Request | TelemetryType.Dependency)]
+        [InlineData(TelemetryType.Metrics | TelemetryType.Request | TelemetryType.Events)]
+        public void CreateFilterWithTracked_WithTelemetryTypeOutsideEnum_Fails(TelemetryType telemetryType)
```
Comment (tomkerkhove) [inline src/Arcus.Observability.Tests.Unit/Telemetry/TelemetryTypeFilterTests.cs:105]: I'm not sure if I get the test case here, can you elaborate please?

File: src/Arcus.Observability.Tests.Unit/Telemetry/TelemetryTypeFilterTests.cs [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+            
+            // Act
+            bool isEnabled = filter.IsEnabled(logEvent);
+            
+            // Assert
+            Assert.False(isEnabled);
+        }
+
+        [Fact]
+        public void LogEventWithoutTelemetry_DoesntFilterAnything_Succeeds()
```
Comment (tomkerkhove) [inline src/Arcus.Observability.Tests.Unit/Telemetry/TelemetryTypeFilterTests.cs:61]: How is this test different from LogEventWithoutValidTelemetry_DoesntFilterAnything_Succeeds?

File: src/Arcus.Observability.Tests.Unit/Telemetry/TelemetryTypeFilterTests.cs [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+            
+            // Act
+            bool isEnabled = filter.IsEnabled(logEvent);
+            
+            // Assert
+            Assert.True(isEnabled);
+        }
+
+        [Fact]
+        public void LogEventAsTelemetry_FiltersOutCorrectTelemetry_Succeeds()
```
Comment (tomkerkhove) [inline src/Arcus.Observability.Tests.Unit/Telemetry/TelemetryTypeFilterTests.cs:39]: I would add a test case like this one, but instead:
1. Tell it to filter a given telemetry type, for example dependency
2. Write a new log event with a different type
3. Entry is not filtered

File: src/Arcus.Observability.Telemetry.Serilog.Filters/TelemetryTypeFilter.cs [hunk 10]
```diff
@@ -44,27 +48,36 @@ public static TelemetryTypeFilter On(TelemetryType telemetryType)
...
             {
                 return true;
             }
 
-            var telemetryType = logEvent.Properties.GetAsEnum<TelemetryType>(ContextProperties.General.TelemetryType);
-            if (telemetryType == null)
+            switch (TelemetryType)
```
Comment (tomkerkhove) [inline src/Arcus.Observability.Telemetry.Serilog.Filters/TelemetryTypeFilter.cs:68]: Can't we revert this change now that we have this?

`ContextProperties.General.TelemetryType`?

Comment (stijnmoreels) [reply src/Arcus.Observability.Telemetry.Serilog.Filters/TelemetryTypeFilter.cs:68]: No because the telemetry type is a property of the log entry type, not the Serilog log event.

File: src/Arcus.Observability.Tests.Unit/Telemetry/TelemetryTypeFilterTests.cs (hunk 8 above)
Comment (stijnmoreels) [reply src/Arcus.Observability.Tests.Unit/Telemetry/TelemetryTypeFilterTests.cs:61]: Woops, yeah, was normally for invalid telemetry contexts but since we can use a more simpler approach, we can remove this one.

File: src/Arcus.Observability.Tests.Unit/Telemetry/TelemetryTypeFilterTests.cs (hunk 9 above)
Comment (stijnmoreels) [reply src/Arcus.Observability.Tests.Unit/Telemetry/TelemetryTypeFilterTests.cs:39]: Yes, missed these ones. I'll add some theory for that. Thanks. Resolving.

File: src/Arcus.Observability.Tests.Unit/Telemetry/TelemetryTypeFilterTests.cs (hunk 7 above)
Comment (stijnmoreels) [reply src/Arcus.Observability.Tests.Unit/Telemetry/TelemetryTypeFilterTests.cs:105]: The telemetry filter guards against misused telemetry type enumeration values. All the inputs must be within the bounds of the enumeration.
This test checks that values outside the scope of the enumeration will result in argument exceptions.

File: src/Arcus.Observability.Telemetry.Core/TelemetryType.cs (hunk 6 above)
Comment (stijnmoreels) [reply src/Arcus.Observability.Telemetry.Core/TelemetryType.cs:7]: Hmm, yeah, could be used outside AI.

Comment (stijnmoreels) [reply src/Arcus.Observability.Telemetry.Core/TelemetryType.cs:7]: Updated this. Thanks!

File: src/Arcus.Observability.Telemetry.Serilog.Filters/TelemetryTypeFilter.cs (hunk 2 above)
Comment (stijnmoreels) [reply src/Arcus.Observability.Telemetry.Serilog.Filters/TelemetryTypeFilter.cs:79]: > When is this method called ?

Serilog filters are predicates on the log level. They're called before the the event is logged to the sink.

> I would assume that it's called to determine if the event must be logged or not ?

Yes.

>  Isnt that done based on the configured loglevel ?

Yes, also, but a filter can be added for extra more grainer control over the events being logged.

> Or am I talking rubbish here ?

Not at all 😉 .

**Review by tomkerkhove — APPROVED** [review]

**Review by fgheysels — APPROVED** [review]
